packages/buidler-core/src/internal/buidler-evm/provider/fork/ForkBlockchain.ts

```diff
@@ -0,0 +1,295 @@
+import Common from "ethereumjs-common";
+import { Transaction } from "ethereumjs-tx";
+import { BN, bufferToInt } from "ethereumjs-util";
+
+import { JsonRpcClient } from "../../jsonrpc/client";
+import {
+  RpcBlockWithTransactions,
+  RpcTransaction,
+  RpcTransactionReceipt,
+} from "../../jsonrpc/types";
+import { BlockchainData } from "../BlockchainData";
+import { FilterParams } from "../node-types";
+import {
+  RpcLogOutput,
+  RpcReceiptOutput,
+  toRpcLogOutput,
+  toRpcReceiptOutput,
+} from "../output";
+import { Block } from "../types/Block";
+import { Blockchain } from "../types/Blockchain";
+import { PBlockchain, toBlockchain } from "../types/PBlockchain";
+
+import { rpcToBlockData } from "./rpcToBlockData";
+import { rpcToTxData } from "./rpcToTxData";
+
+/* tslint:disable only-buidler-error */
+
+export class ForkBlockchain implements PBlockchain {
+  private _data = new BlockchainData();
+  private _latestBlockNumber = this._forkBlockNumber;
+
+  constructor(
+    private _jsonRpcClient: JsonRpcClient,
+    private _forkBlockNumber: BN,
+    private _common: Common
+  ) {}
+
+  public async getLatestBlock(): Promise<Block> {
+    const block = await this.getBlock(this._latestBlockNumber);
+    if (block === undefined) {
+      throw new Error("Block not found");
+    }
+    return block;
+  }
+
+  public async getBlock(
+    blockHashOrNumber: Buffer | number | BN
+  ): Promise<Block | undefined> {
+    if (Buffer.isBuffer(blockHashOrNumber)) {
+      return this._getBlockByHash(blockHashOrNumber);
+    }
+    return this._getBlockByNumber(new BN(blockHashOrNumber));
+  }
+
+  public async addBlock(block: Block): Promise<Block> {
+    const blockNumber = new BN(block.header.number);
+    if (!blockNumber.eq(this._latestBlockNumber.addn(1))) {
+      throw new Error("Invalid block number");
+    }
+    const parent = await this.getLatestBlock();
+    if (!block.header.parentHash.equals(parent.hash())) {
+      throw new Error("Invalid parent hash");
+    }
+    this._latestBlockNumber = this._latestBlockNumber.addn(1);
+    const totalDifficulty = await this._computeTotalDifficulty(block);
+    this._data.addBlock(block, totalDifficulty);
+    return block;
+  }
+
+  public deleteBlock(blockHash: Buffer) {
```

Comment:
I think we kept the `deleteBlock` name (it is `delBlock` in the original VM's `Blockchain`) because deleting a block from a blockchain implies removing all the following blocks as well. 

Other names that come to my mind: `deleteBlocksStartingFrom(block: Block)` or `deleteBlocks(firstBlock: Block)`.

Reply:
no worries